python/hail/methods/statgen.py

```diff
@@ -1677,6 +1677,147 @@ def sample_rows(dataset, fraction, seed=1):
+    implemented as:
+
+    >>> f = functions
+    >>> sm = SplitMulti(ds)
+    >>> pl = f.or_missing(
+    ...      f.is_defined(ds.PL),
+    ...      (f.range(0, 3).map(lambda i: (f.range(0, ds.PL.length())
+    ...                                    .filter(lambda j: f.downcode(f.call(j), sm.a_index()) == f.call(i))
+    ...                                    .map(lambda j: ds.PL[j])
+    ...                                    .min()))))
+    >>> sm.update_rows(a_index=sm.a_index(), was_split=sm.was_split())
+    >>> sm.update_entries(
+    ...     GT=f.downcode(ds.GT, sm.a_index()),
+    ...     AD=f.or_missing(f.is_defined(ds.AD),
+    ...                     [ds.AD.sum() - ds.AD[sm.a_index()], ds.AD[sm.a_index()]]),
+    ...     DP=ds.DP,
+    ...     PL=pl,
+    ...     GQ=f.gq_from_pl(pl))
+    >>> split_ds = sm.result()
+    """
+    @handle_py4j
+    @record_method
+    @typecheck_method(ds=MatrixTable,
+                      keep_star=bool,
+                      left_aligned=bool)
+    def __init__(self, ds, keep_star=False, left_aligned=False):
+        """
+        Parameters
+        ----------
+        ds : :class:`.MatrixTable`
+            An unsplit dataset.
+        keep_star : :obj:`bool`
+            Do not filter out * alleles.
+        left_aligned : :obj:`bool`
+            If ``True``, variants are assumed to be left aligned and have unique
+            loci. This avoids a shuffle. If the assumption is violated, an error
+            is generated.
+
+        Returns
+        -------
+        :class:`.SplitMulti`
+        """
+        self._ds = ds
+        self._keep_star = keep_star
+        self._left_aligned = left_aligned
+        self._entry_fields = None
+        self._row_fields = None
+
+    def new_v(self):
+        """The new, split variant.
+
+        Returns
+        -------
+        :class:`.Expression`
+        """
+        return construct_reference(
+            "newV", type=TVariant(), indices=self._ds._row_indices)
+
+    def a_index(self):
+        """The index of the input allele to the output variant.
+
+        Returns
+        -------
+        :class:`.Int32Expression`
+        """
+        return construct_reference(
+            "aIndex", type=TInt32(), indices=self._ds._row_indices)
+
+    def was_split(self):
+        """``True`` if the original variant was multiallelic.
+
+        Returns
+        -------
+        :class:`.BooleanExpression`
+        """
+        return construct_reference(
+            "wasSplit", type=TBoolean(), indices=self._ds._row_indices)
+
+    def update_rows(self, **kwargs):
+        """Set the row field updates for this SplitMulti object.
+
+        Note
+        ----
+        May only be called once.
+        """
+        if self._row_fields is None:
+            self._row_fields = kwargs
+        else:
+            raise FatalError("You may only call update_rows once")
+
+    def update_entries(self, **kwargs):
```

Comment:
why are these not `annotate`? that's consistent with filter alleles. Why should we introduce a new term?

Reply:
It can only be called once on this object, so it breaks the spirit of annotate. I purposely want the user to be confused and read the docs on how this works.

I don't know how to make repeated calls work because the user (reasonably) would expect to use previously annotated fields. To make that work I'd have to let-bind the values so they can be used to set both fields. I can't put a let outside the field assignment.